packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts

```diff
@@ -28,26 +26,120 @@ const protocolVersions = ["^0.4.0", "^0.3.0", "^0.2.0", "^0.1.0"];
 const socketReferenceBufferTime = 2000;
 
 class SocketReference {
-    public references: number = 1;
-    public delayDeleteTimeout?: NodeJS.Timeout;
-    public delayDeleteTimeoutSetTime?: number;
+    private references: number = 1;
+    private delayDeleteTimeout?: NodeJS.Timeout;
+    private delayDeleteTimeoutSetTime?: number;
+    private _socket: SocketIOClient.Socket | undefined;
+    private inUse = false;
 
-    public constructor(public socket: SocketIOClient.Socket | undefined) {
+    // Map of all existing socket io sockets. [url, tenantId, documentId] -> socket
+    private static readonly socketIoSockets: Map<string, SocketReference> = new Map();
+
+    public static find(key: string, logger: ITelemetryLogger) {
+        const socketReference = SocketReference.socketIoSockets.get(key);
+
+        // Verify the socket is healthy before reusing it
+        if (socketReference && socketReference.disconnected) {
+            // The socket is in a bad state. fully remove the reference
+            socketReference.closeSocket();
+            return undefined;
+        }
+
+        if (socketReference) {
+            // Clear the pending deletion if there is one
+            socketReference.clearTimer();
+            socketReference.references++;
+
+            logger.sendTelemetryEvent({
+                references: socketReference.references,
+                eventName: "OdspDocumentDeltaCollection.GetSocketIoReference",
+                delayDeleteDelta: socketReference.delayDeleteTimeoutSetTime !== undefined ?
+                    (Date.now() - socketReference.delayDeleteTimeoutSetTime) : undefined,
+            });
+        }
+
+        return socketReference;
+    }
+
+    /**
+     * Removes a reference for the given key
+     * Once the ref count hits 0, the socket is disconnected and removed
+     * @param key - socket reference key
+     * @param isFatalError - true if the socket reference should be removed immediately due to a fatal error
+     */
+    public removeSocketIoReference(isFatalError: boolean) {
+        assert(this.references > 0);
+        this.references--;
+
+        // see comment in disconnected() getter
+        this.inUse = true;
```

Comment:
This feels like a strange place to set the flag (though maybe the naming is throwing me, that we're saying it's "in use" when the reference gets removed feels unintuitive).  Seems like this wants to get set when the connection completes maybe?  Or otherwise just the name change?

Reply:
Change name to isPendingInitialConnection and reversed meaning (true/false).
It's a bit more than successful connection, it tracks any attempt to connection (including failed connections).
We can add finally block around initialize() call, but I really want to keep all socket reuse logic in this class and have as little leakage as possible.

Listening for "connected" event on socket would likely be sufficient, but I want to ensure that even in case of unsuccessful connection (like timeout) we do not reuse socket. I've changed comment to in disconnected() to better reflect all cases.